### PR TITLE
JBR-7194: extension-based filters in native file dialogs

### PR DESCRIPTION
**Pull request by trespasserw**
Adding an ability to set simple file-extension-based filters for file dialogs. While being less flexible than `FileDialog#setFilenameFilter`, extension-based filters don't require callbacks to Java code and are supported on all platforms.

### CODE REVIEW

**Comment by trespasserw** [comment]
(WIP; the implementation for macOS comes later)

**Comment by avu** [comment]
@trespasserw It would be nice to have a regression test for this feature (something like test/jdk/jb/javax/swing/CustomTitleBar/JFrameNativeControlsMacOSTest.java). It can be created even for incomplete implementation using the platform tag in the jtreg test case.

**Comment by vprovodin** [comment]
Should the ticket IJPL-150176 be moved into the JBR YT project?

**Comment by trespasserw** [comment]
@vprovodin what are those commits?

IJPL-150176 should not be moved, but I will create a companion ticket in JBR later today and edit commit messages.

File: src/java.desktop/share/classes/com/jetbrains/desktop/JBRFileDialog.java [hunk 1]
```diff
@@ -55,9 +42,14 @@ public int getHints() {
         return hints;
     }
 
-    public void setLocalizationStrings(String openButtonText, String selectFolderButtonText) {
+    public void setLocalizationStrings(String openButtonText, String selectFolderButtonText, String allFilesFilterDescription) {
```
Comment (YaaZ) [inline src/java.desktop/share/classes/com/jetbrains/desktop/JBRFileDialog.java:45]: Removing an API method is a breaking change, I think adding more generic `setLocalizationString(key, value)` would be better

File: src/java.desktop/share/classes/com/jetbrains/desktop/JBRFileDialog.java [hunk 2]
```diff
@@ -25,28 +25,15 @@ public static JBRFileDialog get(FileDialog dialog) {
...
-    /**
-     * Text for "Open" button (used when common file dialogs are enabled on
-     * Windows).
-     */
     public String openButtonText;
-
-    /**
-     * Text for "Select Folder" button (used when common file dialogs are
-     * enabled on Windows).
-     */
```
Comment (YaaZ) [inline src/java.desktop/share/classes/com/jetbrains/desktop/JBRFileDialog.java:48]: What is the purpose of removing this javadoc?

File: src/jetbrains.api/src/com/jetbrains/JBRFileDialog.java [hunk 3]
```diff
@@ -37,12 +42,11 @@ static JBRFileDialog get(FileDialog dialog) {
...
     /**
-     * Set JBR-specific file dialog hints:
+     * Set file dialog hints:
      * <ul>
-     *     <li>SELECT_FILES_HINT, SELECT_DIRECTORIES_HINT - Whether to select files, directories or both
-     *     (used when common file dialogs are enabled on Windows, or on macOS),
-     *     both unset bits are treated as a default platform-specific behavior</li>
-     *     <li>CREATE_DIRECTORIES_HINT - Whether to allow creating directories or not (used on macOS)</li>
+     *     <li>SELECT_FILES_HINT, SELECT_DIRECTORIES_HINT - Whether to select files, directories, or both;
+     *     neither bit set is treated as a default platform-specific behavior</li>
```
Comment (YaaZ) [inline src/jetbrains.api/src/com/jetbrains/JBRFileDialog.java:48]: I think original wording was more clear and new one is kinda ambiguous:
* "both unset bits are treated as a default platform-specific behavior" -> if neither of `SELECT_FILES_HINT` or `SELECT_DIRECTORIES_HINT` is set, then behavior is platform-specific
* "neither bit set is treated as a default platform-specific behavior" -> No combination of bits will result in platform-specific behavior?

**Review by YaaZ — CHANGES_REQUESTED** [review]

File: src/java.desktop/share/classes/com/jetbrains/desktop/JBRFileDialog.java (hunk 1 above)
Comment (trespasserw) [reply src/java.desktop/share/classes/com/jetbrains/desktop/JBRFileDialog.java:45]: AFAIU, classes in `com.jetbrains.desktop` are part of the corresponding JBR module which the client code is not supposed to use directly. It is proxied by `com.jetbrains.JBRFileDialog` in `jbr-api`, which preserves compatibility.

A key-based approach is an interesting idea; I'll try.

File: src/java.desktop/share/classes/com/jetbrains/desktop/JBRFileDialog.java (hunk 2 above)
Comment (trespasserw) [reply src/java.desktop/share/classes/com/jetbrains/desktop/JBRFileDialog.java:48]: The comments weren't removed, they were moved from a "backend" class to the "API" one.

File: src/jetbrains.api/src/com/jetbrains/JBRFileDialog.java (hunk 3 above)
Comment (trespasserw) [reply src/jetbrains.api/src/com/jetbrains/JBRFileDialog.java]: Updated.

File: src/java.desktop/share/classes/com/jetbrains/desktop/JBRFileDialog.java (hunk 1 above)
Comment (YaaZ) [reply src/java.desktop/share/classes/com/jetbrains/desktop/JBRFileDialog.java:45]: That's correct, but as you added default implementation for this method in`jbr-api`, means that just running on a newer JBR (without updating jbr-api dependency) would already be incompatible, as there would be no implementation for that API method. I suggest leaving this method as is, we can just mark is as deprecated for the time being

**Comment by YaaZ** [comment]
Pushed to `main` for now

**Comment by YaaZ** [comment]
Pushed this to `main`, `jbr17`, `jbr21`, so closing the review

**Pull request closed without merging**